### PR TITLE
Extended Keyword Parsing to cover member lookup expressions.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Fix for #29231
Extend the parsing for keywords, to look ahead at the next token to ee if it is a qualifier token.

If it is then treat the keyword as an identifier (with an attached diagnostic) thus considering it member lookup expression.
This allows Interpolation String to "correctly" parse this form as an expression, thus the rest is parsed as part of the string interpolation and not treat the expression as text.

keywordexpr_01

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 1]
```diff
@@ -9,7 +9,7 @@ Imports InternalSyntaxFactory = Microsoft.CodeAnalysis.VisualBasic.Syntax.Intern
 
 Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
 
-    Friend Partial Class Parser
+    Partial Friend Class Parser
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: Please revert

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 2]
```diff
@@ -46,7 +46,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
         '    bool BailIfFirstTokenRejected                 // bail (return NULL) if the first token isn't a valid expression-starter, rather than reporting an error or setting ErrorInConstruct
         Private Function ParseExpressionCore(
             Optional pendingPrecedence As OperatorPrecedence = OperatorPrecedence.PrecedenceNone,
-            Optional bailIfFirstTokenRejected As Boolean = False
+            Optional bailIfFirstTokenRejected As Boolean = False,
+            Optional PrevTerm As ExpressionSyntax = Nothing
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: Please use camelCased names.

File: src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb [hunk 3]
```diff
@@ -10,10 +10,10 @@ Imports InternalSyntaxFactory = Microsoft.CodeAnalysis.VisualBasic.Syntax.Intern
 
 Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
 
-    Friend Partial Class Parser
+    Partial Friend Class Parser
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb]: Please revert.

**Comment by jaredpar** [comment]
This PR does not follow our coding style in pretty much every change. Please correct this.

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 4]
```diff
@@ -421,7 +421,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
+                            Dim _KeywordAsIdentifier_ = ParseSimpleNameExpressionAllowingKeywordAndTypeArguments()
+                            _KeywordAsIdentifier_ = ReportSyntaxError(_KeywordAsIdentifier_, ERRID.ERR_InvalidUseOfKeyword)
+                            Dim _InvocationExpression_ = ParseExpressionCore(PrevTerm:=_KeywordAsIdentifier_)
+                            If _InvocationExpression_.Kind = SyntaxKind.InvocationExpression Then
+                                Return _InvocationExpression_
+                            End If
+                        End If
+                        GoTo MissingExpression
                     Else
+MissingExpression:
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: a goto here seems unnecessary.  why not just replace the else with an "End If"?

File: src/Compilers/VisualBasic/Test/Syntax/Parser/InterpolatedStringParsingTests.vb [hunk 5]
```diff
@@ -922,4 +922,22 @@ BC30648: String constants must end with a double quote.
...
+  Public Overrides Function ToString() As String
+    Return $""{End.ToString()})""
+  End Function
+End Class"
+        Dim result = Parse(code)
+        Dim diagnostics = result.GetDiagnostics.ToImmutableListOrEmpty
+        result.AssertTheseDiagnostics(
+        <expected>
+BC30183: Keyword is not valid as an identifier.
+    Return $"{End.ToString()})"
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Test/Syntax/Parser/InterpolatedStringParsingTests.vb]: can we have a lot more tests than this?  we'll need tests of keywords in many different areas followed by any allowable delimiter.  We should also have specific tree and diagnostics tests

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb (hunk 1 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: @jaredpar  I think this is caused by an something in the IDE, possible pretty lister? Or Code Style enforcement? 
Writing `Friend Partial` is converted to `Partial Friend` automatically.

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 6]
```diff
@@ -410,19 +412,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
+                            If nextToken.Kind = SyntaxKind.DotToken OrElse nextToken.Kind = SyntaxKind.QuestionToken Then
+                                Dim keywordAsIdentifier = ParseSimpleNameExpressionAllowingKeywordAndTypeArguments()
+                                keywordAsIdentifier = ReportSyntaxError(keywordAsIdentifier, ERRID.ERR_InvalidUseOfKeyword)
+                                Dim _InvocationExpression_ = ParseExpressionCore(prevTerm:=keywordAsIdentifier)
+                                If _InvocationExpression_.Kind = SyntaxKind.InvocationExpression Then
+                                    Return _InvocationExpression_
+                                End If
+                            End If
+
+                        End If
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: extra newline unnecessary. though you can potentially put a comment there saying it is intentional that you are falling through.

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 7]
```diff
@@ -410,19 +412,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
-                        If BailIfFirstTokenRejected Then
+                        If start.IsKeyword Then
+
+                            ' Is it in the form : [keyword] [qualifier] [identifier]
+                            ' If so treat the keyword as an identifier and continue to parse.
+                            Dim nextToken = PeekNextToken()
+                            If nextToken.Kind = SyntaxKind.DotToken OrElse nextToken.Kind = SyntaxKind.QuestionToken Then
+                                Dim keywordAsIdentifier = ParseSimpleNameExpressionAllowingKeywordAndTypeArguments()
+                                keywordAsIdentifier = ReportSyntaxError(keywordAsIdentifier, ERRID.ERR_InvalidUseOfKeyword)
+                                Dim _InvocationExpression_ = ParseExpressionCore(prevTerm:=keywordAsIdentifier)
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: _Foo_ is not our naming pattern for locals.

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 8]
```diff
@@ -410,19 +412,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
+                                If _InvocationExpression_.Kind = SyntaxKind.InvocationExpression Then
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: why are you only allowing invocation expressions.  this won't help if you did something like ```$"{End.Prop}"```

File: src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb [hunk 9]
```diff
@@ -332,4 +332,4 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
 
     End Class
 
-End Namespace
+End Namespace
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb]: please undo this.

File: src/Compilers/VisualBasic/Test/Syntax/Parser/InterpolatedStringParsingTests.vb (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/VisualBasic/Test/Syntax/Parser/InterpolatedStringParsingTests.vb]: this also affects far more than interpolated strings.  i would expect a lot of statement/expression parsing tests.

File: src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb (hunk 3 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseInterpolatedString.vb]: See Issue #27893

**Comment by CyrusNajmabadi** [comment]
@AdamSpeight2008 Please do not 'format document' in the compiler codebase.  If/when the compiler wants to format things, they would do it as an entirely independent PR.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi  Formatting changes are partially due to Issue #27893. 
Are there any additional unit tests that I could add?

**Comment by CyrusNajmabadi** [comment]
@AdamSpeight2008 Your change is unrelated to interpolated string parsing.  I would keep around the tests you have.  But you need more general tests showing a variety of keywords followed by delimiters in a variety of different circumstances.  

I would check with other compiler devs to see waht else they would want.  If this was C# i know it would be good to have tests that validate the tree shape.  a-la: https://github.com/dotnet/roslyn/blob/4cd7d5efd812fbb39bcf03d0b45bf2d4f13ab0ba/src/Compilers/CSharp/Test/Syntax/Parsing/AsyncParsingTests.cs#L42-L49

But i don't know if VB has anything equivalent.

We'd likely also want some semantic and IOperation tests to make sure this doesn't cause anything downstream to crash.

**Comment by jaredpar** [comment]
Closing as this PR has gotten stale at this point.

**Pull request closed without merging**